pandas/core/categorical.py

```diff
@@ -1027,19 +1027,25 @@ def value_counts(self, dropna=True):
         """
         import pandas.hashtable as htable
         from pandas.core.series import Series
-
+        from pandas.core.index import CategoricalIndex
         cat = self.dropna() if dropna else self
         keys, counts = htable.value_count_int64(com._ensure_int64(cat._codes))
-        result = Series(counts, index=keys)
 
-        ix = np.arange(len(cat.categories), dtype='int64')
-        if not dropna and -1 in keys:
-            ix = np.append(ix, -1)
-        result = result.reindex(ix, fill_value=0)
-        result.index = (np.append(cat.categories, np.nan)
```

Comment:
Doesn't quite work since apparently if `np.nan` is in `.categories` then the code isn't always -1 (which seems like a bug to me?)